### PR TITLE
Stop generating parent-defined alias methods in child classes

### PR DESCRIPTION
**Pull request by ipc103**
### Motivation / Background

We currently issue a deprecation warning when an attribute method is overwritten to let people know that they'll need to manually define the alias override as well. If the override is also defined on the parent class, then we shouldn't need to issue a deprecation warning. 

Before, the following was raising a deprecation warning for because `subject_was` is not implemented directly on the child class.

```ruby
class Topic < ActiveRecord::Base
  alias_attribute :subject, :title
  
  def title_was
    "override_title"
  end
  alias_method :subject_was, :title_was
end

class SubTopic < Topic
end
```

Following this change, the behavior works correctly without a deprecation warning.

### Detail

Because we're using a class_attribute to track all of the attribute_aliases, each subclass was regenerating the parent class methods. When a child was generating methods for the parent, it was only looking to see if the manual override method was defined within. 

Instead, we can use a class instance variable to track the attributes aliased locally in that particular subclass. We then walk up the chain and re-define the attribute methods if they haven't been defined yet.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Review by nvasilevski — COMMENTED** [review]
Hey, thank you for the report & the fix! 

I just wanted to provide a little bit context on the problem. If we look at the deprecation generated by the test without fix:
```
DEPRECATION WARNING: AttributeMethodsTest::ChildWithDeprecatedBehaviorResolved model aliases `title` and has a method called `title_was` defined.
```
You may notice that it's actually misleading as it says that the child class has both alias and manual definition and neither of it is true. 

So a lower-level cause for this problem is the fact that the child class is trying to redefine parent's aliases. It's a known problem but we decided to address it later since it shouldn't have had any impact on the application. We didn't consider the deprecation being wrongly issued from a child class. 

So ideally I would prefer this issue to be addressed by Rails not attempting to generate parent-defined aliases in child classes. The first solution that comes to mind is to have another storage parallel to the existing `attribute_aliases` 
https://github.com/rails/rails/blob/8ec27a4529e37cff77abe54573bb8923d9fe8e9a/activemodel/lib/active_model/attribute_methods.rb#L205
and the new one will only keep track of the current class aliases so we can use it to generate alias attribute methods instead of the one that keeps track of all aliases in the inheritance chain 

https://github.com/rails/rails/blob/8ec27a4529e37cff77abe54573bb8923d9fe8e9a/activerecord/lib/active_record/attribute_methods.rb#L73

File: activerecord/test/cases/attribute_methods_test.rb
Comment (nvasilevski) [inline activerecord/test/cases/attribute_methods_test.rb:1241]: It made me realized that we end up using wrong deprecator - https://github.com/rails/rails/pull/48917/files

**Comment by ipc103** [comment]
> So ideally I would prefer this issue to be addressed by Rails not attempting to generate parent-defined aliases in child classes. The first solution that comes to mind is to have another storage parallel to the existing attribute_aliases

Thanks! That makes sense to me and I agree, that sounds like a better fix. I'll push an update to this branch for review 🙏

File: activemodel/lib/active_model/attribute_methods.rb
Comment (ipc103) [inline activemodel/lib/active_model/attribute_methods.rb:365]: There is probably a better name for this, but this was the best I could come up for now.

**Comment by ipc103** [comment]
@nvasilevski Thanks again for the feedback/idea - I made those changes in 6c62446. Let me know if you think this is a better approach or if you have any better ideas around naming than I did 😬

File: activerecord/lib/active_record/attribute_methods.rb
Comment (nvasilevski) [inline activerecord/lib/active_record/attribute_methods.rb:70]: It looks like we want to generate parent's alias attribute regardless of whether `@alias_attributes_mass_generated` is true or not. Should we just always call it at the first place like:
```ruby
klass.superclass.generate_alias_attributes
return if @alias_attributes_mass_generated
```
so we won't need to call it in the `synchronize` block?

File: activerecord/lib/active_record/attribute_methods.rb
Comment (nvasilevski) [inline activerecord/lib/active_record/attribute_methods.rb:69]: Do we want to check for `base_class?` instead?
```suggestion
        return if base_class?
```

Similarly to Do we want to use 
```suggestion
        return if base_class?
```

https://github.com/rails/rails/blob/a0b216a8a5eab5ebb0190aebcf3628c0ae5f5399/activerecord/lib/active_record/attribute_methods.rb#L117

File: activerecord/lib/active_record/attribute_methods.rb
Comment (ipc103) [reply activerecord/lib/active_record/attribute_methods.rb]: Good idea - updated in 217161c

**Review by nvasilevski — APPROVED** [review]
Thank you!

**Comment by nvasilevski** [comment]
@ipc103 Would you mind squashing commits into one? Core team generally asks for that before merging

I'll ask on Discord for someone from the core team to have a look

**Comment by ipc103** [comment]
Thanks @nvasilevski ! Sorry for the extra labels [here](https://github.com/rails/rails/pull/48913#event-10113165152), my rebase went awry 😅 Should be good now I think 🤞

**Comment by nvasilevski** [comment]
@ipc103 Sorry one last thing that I missed, could you change the PR title to be relevant to the changes as we pivoted from changing deprecation to changing the generation logic instead. Thank you!

**Comment by ipc103** [comment]
@nvasilevski Sure thing - sorry I didn't see your message before this merged, but I updated anyway for posterity. 

Thanks everyone for helping here - this was my first PR to rails, so I especially appreciate the guidance and support ❤️

**Pull request merged**
